pkg/connection/connection.go

```diff
@@ -0,0 +1,420 @@
+		NumRows:         len(data[0]),
+		Data:            data,
+		Attributes: types.Attributes{
+			ResultSetMaxRows: c.Config.ResultSetMaxRows,
+		},
+	}
+	result := &types.SqlQueriesResponse{}
+	err := c.Send(ctx, command, result)
+	if err != nil {
+		return nil, err
+	}
+	if result.NumResults == 0 {
+		return nil, errors.ErrMalformedData
+	}
+	return result, c.closePreparedStatement(ctx, s)
+}
+
+func (c *Connection) closePreparedStatement(ctx context.Context, s *types.CreatePreparedStatementResponse) error {
+	return c.Send(ctx, &types.ClosePreparedStatementCommand{
+		Command:         types.Command{Command: "closePreparedStatement"},
+		StatementHandle: s.StatementHandle,
+	}, nil)
+}
+
+func (c *Connection) exec(ctx context.Context, query string, args []driver.Value) (driver.Result, error) {
+	if c.IsClosed {
+		logger.ErrorLogger.Print(errors.ErrClosed)
+		return nil, driver.ErrBadConn
+	}
+	result := make(chan driver.Result, 1)
+	errs, errctx := errgroup.WithContext(ctx)
+
+	if utils.IsImportQuery(query) {
+		originalQuery := query
+		p, err := c.getProxy()
+		if err != nil {
+			return nil, err
+		}
+		err = p.StartProxy()
+		if err != nil {
+			return nil, err
+		}
+		defer p.Close()
+		query = utils.UpdateImportQuery(originalQuery, p.Host, p.Port)
+		errs.Go(func() error { return c.uploadFiles(errctx, p, originalQuery) })
+	}
+	// No values provided, simple execute is enough
+	if len(args) == 0 {
+		errs.Go(c.executeSimpleWrapper(errctx, query, result))
+	} else {
+		errs.Go(c.executePreparedStatementWrapper(errctx, query, args, result))
+	}
+	err := errs.Wait()
+	close(result)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return <-result, nil
+}
+
+func (c *Connection) executeSimpleWrapper(ctx context.Context, query string, result chan driver.Result) func() error {
+	return func() error {
+		r, err := c.executeSimpleWithResult(ctx, query)
+		if err != nil {
+			return err
+		}
+		result <- r
+		return nil
+	}
+}
+
+func (c *Connection) executePreparedStatementWrapper(ctx context.Context, query string, args []driver.Value, result chan driver.Result) func() error {
+	return func() error {
+		prepResponse := &types.CreatePreparedStatementResponse{}
+		err := c.Send(ctx, &types.CreatePreparedStatementCommand{
+			Command: types.Command{Command: "createPreparedStatement"},
+			SQLText: query,
+		}, prepResponse)
+		if err != nil {
+			return err
+		}
+		resp, err := c.executePreparedStatement(ctx, prepResponse, args)
+		if err != nil {
+			return err
+		}
+		r, err := ToResult(resp)
+		if err != nil {
+			return err
+		}
+		result <- r
+		return nil
+	}
+}
+
+func (c *Connection) getProxy() (*proxy.Proxy, error) {
+	hosts, err := utils.ResolveHosts(c.Config.Host)
+	if err != nil {
+		return nil, err
+	}
+	utils.ShuffleHosts(hosts)
+	return proxy.NewProxy(hosts, c.Config.Port)
+}
+
+func (c *Connection) uploadFiles(ctx context.Context, p *proxy.Proxy, query string) error {
+	paths, err := utils.GetFilePaths(query)
```

Comment:
Idea for refactoring (not right now): create a new struct for `IMPORT` queries with all necessary parts (file paths, row separator, ...). Then you parse it only once and don't need to extract infos every time you need some info from it.